File: api/budget_kind_engine.py
"""Deterministic B/L/F/S/U/Z budget calculation engine with trace output."""

from __future__ import annotations

from dataclasses import dataclass, asdict
from decimal import Decimal
from typing import Any

from api.decimal_math import multiply, quantize, sum_values


@dataclass(frozen=True)
class TraceStep:
    operation: str
    inputs: dict[str, Any]
    result: str


@dataclass(frozen=True)
class CalculationTrace:
    kind: str
    scale: int
    steps: list[TraceStep]
    result: str

    def to_dict(self) -> dict[str, Any]:
        return {
            "kind": self.kind,
            "scale": self.scale,
            "steps": [asdict(step) for step in self.steps],
            "result": self.result,
        }


def _decimal(value: Any) -> Decimal:
    return Decimal(str(value))


def calculate_budget_kind(kind: str, payload: dict[str, Any], scale: int) -> CalculationTrace:
    normalized = kind.upper().strip()
    if scale < 0 or scale > 8:
        raise ValueError("scale must be between 0 and 8")
    steps: list[TraceStep] = []

    if normalized == "L":
        quantity = str(payload.get("quantity", "0"))
        unit_price = str(payload.get("unit_price", "0"))
        result = multiply(quantity, unit_price, scale)
        steps.append(TraceStep("MULTIPLY", {"quantity": quantity, "unit_price": unit_price}, result))
    elif normalized in {"B", "Z"}:
        children = [str(value) for value in payload.get("children", [])]
        result = sum_values(children, scale)
        steps.append(TraceStep("SUM_CHILDREN", {"children": children}, result))
    elif normalized == "F":
        base = str(payload.get("base", "0"))
        rate = str(payload.get("rate", "0"))
        result = multiply(base, rate, scale)
        steps.append(TraceStep("MULTIPLY_BASE_RATE", {"base": base, "rate": rate}, result))
    elif normalized == "S":
        remaining = _decimal(payload.get("base", "0"))
        if remaining < 0:
            raise ValueError("tiered base cannot be negative")
        previous_limit = Decimal("0")
        subtotal = Decimal("0")
        tiers = payload.get("tiers", [])
        if not isinstance(tiers, list) or not tiers:
            raise ValueError("tiers are required for S items")
        for tier in tiers:
            rate = _decimal(tier.get("rate", "0"))
            up_to_raw = tier.get("up_to")
            if up_to_raw is None:
                quantity = remaining
            else:
                up_to = _decimal(up_to_raw)
                if up_to < previous_limit:
                    raise ValueError("tiers must be ordered by up_to")
                capacity = up_to - previous_limit
                quantity = min(remaining, capacity)
                previous_limit = up_to
            tier_amount = quantity * rate
            subtotal += tier_amount
            step_result = quantize(str(tier_amount), scale)
            steps.append(TraceStep("TIER", {"quantity": str(quantity), "rate": str(rate), "up_to": up_to_raw}, step_result))
            remaining -= quantity
            if remaining <= 0:
                break
        if remaining > 0:
            raise ValueError("tier schedule does not cover base")
        result = quantize(str(subtotal), scale)
    elif normalized == "U":
        signed_values: list[str] = []
        terms = payload.get("terms", [])
        for term in terms:
            sign = int(term.get("sign", 1))
            if sign not in {-1, 1}:
                raise ValueError("term sign must be -1 or 1")
            signed_values.append(str(_decimal(term.get("amount", "0")) * sign))
        result = sum_values(signed_values, scale)
        steps.append(TraceStep("SIGNED_SUM", {"terms": terms}, result))
    else:
        raise ValueError(f"unsupported budget item kind: {kind}")

    return CalculationTrace(normalized, scale, steps, result)

File: pcces-go/internal/domain/money/budget_kind_engine.go
package money

import (
	"fmt"
	"math/big"
	"strings"
)

type BudgetTerm struct { Sign int `json:"sign"`; Amount string `json:"amount"` }
type BudgetTier struct { UpTo *string `json:"up_to,omitempty"`; Rate string `json:"rate"` }
type BudgetKindInput struct {
	Quantity string `json:"quantity,omitempty"`; UnitPrice string `json:"unit_price,omitempty"`
	Children []string `json:"children,omitempty"`; Base string `json:"base,omitempty"`; Rate string `json:"rate,omitempty"`
	Tiers []BudgetTier `json:"tiers,omitempty"`; Terms []BudgetTerm `json:"terms,omitempty"`
}
type BudgetTraceStep struct { Operation string `json:"operation"`; Inputs map[string]any `json:"inputs"`; Result string `json:"result"` }
type BudgetCalculationTrace struct { Kind string `json:"kind"`; Scale int `json:"scale"`; Steps []BudgetTraceStep `json:"steps"`; Result string `json:"result"` }

func rat(value string) (*big.Rat,error){ r,ok:=new(big.Rat).SetString(value); if !ok{return nil,fmt.Errorf("invalid decimal %q",value)}; return r,nil }
func minRat(a,b *big.Rat)*big.Rat{if a.Cmp(b)<=0{return new(big.Rat).Set(a)};return new(big.Rat).Set(b)}

func CalculateBudgetKind(kind string, in BudgetKindInput, scale int)(BudgetCalculationTrace,error){
	k:=strings.ToUpper(strings.TrimSpace(kind)); trace:=BudgetCalculationTrace{Kind:k,Scale:scale,Steps:[]BudgetTraceStep{}}
	var result string; var err error
	switch k {
	case "L": result,err=Multiply(in.Quantity,in.UnitPrice,scale); trace.Steps=append(trace.Steps,BudgetTraceStep{"MULTIPLY",map[string]any{"quantity":in.Quantity,"unit_price":in.UnitPrice},result})
	case "B","Z": result,err=Sum(in.Children,scale); trace.Steps=append(trace.Steps,BudgetTraceStep{"SUM_CHILDREN",map[string]any{"children":in.Children},result})
	case "F": result,err=Multiply(in.Base,in.Rate,scale); trace.Steps=append(trace.Steps,BudgetTraceStep{"MULTIPLY_BASE_RATE",map[string]any{"base":in.Base,"rate":in.Rate},result})
	case "U": vals:=make([]string,0,len(in.Terms)); for _,t:=range in.Terms{if t.Sign!=1&&t.Sign!=-1{return trace,fmt.Errorf("term sign must be -1 or 1")}; if t.Sign<0{vals=append(vals,"-"+strings.TrimPrefix(t.Amount,"+"))}else{vals=append(vals,t.Amount)}}; result,err=Sum(vals,scale); trace.Steps=append(trace.Steps,BudgetTraceStep{"SIGNED_SUM",map[string]any{"terms":in.Terms},result})
	case "S":
		remaining,e:=rat(in.Base); if e!=nil{return trace,e}; if remaining.Sign()<0{return trace,fmt.Errorf("tiered base cannot be negative")}; previous:=new(big.Rat); subtotal:=new(big.Rat)
		if len(in.Tiers)==0{return trace,fmt.Errorf("tiers are required for S items")}
		for _,tier:=range in.Tiers{rate,e:=rat(tier.Rate);if e!=nil{return trace,e};qty:=new(big.Rat).Set(remaining);var up any
			if tier.UpTo!=nil{limit,e:=rat(*tier.UpTo);if e!=nil{return trace,e};if limit.Cmp(previous)<0{return trace,fmt.Errorf("tiers must be ordered")};capacity:=new(big.Rat).Sub(limit,previous);qty=minRat(remaining,capacity);previous.Set(limit);up=*tier.UpTo}
			amount:=new(big.Rat).Mul(qty,rate);subtotal.Add(subtotal,amount);stepResult,_:=Quantize(amount.FloatString(12),scale);trace.Steps=append(trace.Steps,BudgetTraceStep{"TIER",map[string]any{"quantity":qty.FloatString(12),"rate":tier.Rate,"up_to":up},stepResult});remaining.Sub(remaining,qty);if remaining.Sign()<=0{break}}
		if remaining.Sign()>0{return trace,fmt.Errorf("tier schedule does not cover base")};result,err=Quantize(subtotal.FloatString(12),scale)
	default:return trace,fmt.Errorf("unsupported budget item kind: %s",kind)
	}
	if err!=nil{return trace,err};trace.Result=result;return trace,nil
}
